File: Problem2_MeetingRoomsii.java
// Time Complexity : o(NlogN)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        
        if(intervals==null || intervals.length==0)
            return 0;
        
        int minCount=1;
         Arrays.sort(intervals,(a,b)->(a[0]-b[0]));
        
        PriorityQueue<Integer> pq=new PriorityQueue();
        pq.add(intervals[0][1]);
        
       
             
                    for(int i=1;i<intervals.length;i++)
                    {
                
                            if(pq.peek() <= intervals[i][0])
                            {
                                pq.poll();
                            }
                            else
                            {
                                minCount++;
                            }
                            pq.add(intervals[i][1]);
                        }
                    return minCount;
                    
}
    }
                    

File: Problem2_MeetingRoomsII.py
# Time Complexity: O(n logn), where n - length of the intervals
# Space Complexity: O(n)

class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals or len(intervals) == 0:
            return 0

        # Sort the intervals based on the start time
        intervals.sort(key=lambda x: x[0])

        # Store the end time of intervals in a Heap
        heap = []
        heapq.heappush(heap, intervals[0][1])

        # Compare the start time of every interval with the minimum end time (peek val)
        for i in range(1, len(intervals)):
            start = intervals[i][0]
            end = intervals[i][1]

            # Use the same room if it starts later the minimum end time, else add a new room
            if start >= heap[0]:
                heapq.heappop(heap)
            heapq.heappush(heap, end)

        return len(heap)

